examples/clusterclass/class/class.yaml

```diff
@@ -0,0 +1,372 @@
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3ClusterTemplate
+metadata:
+  name: example-cluster-template
+  namespace: ${NAMESPACE}
+spec:
+  template:
+    spec:
+      controlPlaneEndpoint:
+        host: ${CLUSTER_APIENDPOINT_HOST}
+        port: ${CLUSTER_APIENDPOINT_PORT}
```

Comment:
Done. Added variables and patches to `ClusterClass` and referenced those from `Cluster` itself. Seems to work fine regarding to my tests.
Just one question - currently `Metal3ClusterTemplate` is [validated by the webhook for `controlPlaneEndpoint` definition](https://github.com/p-strusiewiczsurmacki-mobica/cluster-api-provider-metal3/blob/clusterclass-support/api/v1beta1/metal3clustertemplate_webhook.go#L65), meaning `something` has always be defined there. Should this validation be omitted and we should allow empty definition (without specifying controlPlaneEndpoint`)?

Reply:
Not sure I understood the question. I think the controlPlaneEndpoint is always required. How would the cluster work without it?

Reply:
Currently `Metal3ClusterTemplate` is required to have `controlPlaneEndpoint` defined, e.g:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3ClusterTemplate
metadata:
  name: example-cluster-template
spec:
  template:
    spec:
      controlPlaneEndpoint:
        host: 127.0.0.1
        port: 6443
      noCloudProvider: true
```

But I think that makes it not reusable, right? Do we want to accept something like this (with no `controlPlaneEndpoint`):

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3ClusterTemplate
metadata:
  name: example-cluster-template
spec:
  template:
    spec:
      noCloudProvider: true
```

And then controlPlaneEndpoint should be defined in `Cluster` resource, I think?

Reply:
Ok now I understand!
I think the way to handle this is to use variables in the ClusterClass. So set it like you already did in the Metal3ClusterTemplate, but then add a variable and patch to override it. Then it would be possible to create clusters like

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: my-cluster
spec:
  ...
  topology:
    class: example-clusterclass
    version: v1.29.0
    variables:
    - name: controlPlaneEndpointHost
      value: 192.168.0.100
```